### PR TITLE
Fix closure compiler errors in JS libraries. NFC

### PR DESCRIPTION
**Pull request by sbc100**
Expanding `test_closure_full_js_library` exposed a whole
bunch of bugs in our JS code.

### CODE REVIEW

File: src/library_exceptions.js [hunk 1]
```diff
@@ -110,6 +110,7 @@ var LibraryExceptions = {
     };
   },
 
+#if !DISABLE_EXCEPTION_CATCHING
```
Comment (kripken) [inline src/library_exceptions.js:113]: what error happens without this change?

Comment (sbc100) [reply src/library_exceptions.js:113]: Undefined symbol `__cxa_is_pointer_type` below.

Comment (sbc100) [reply src/library_exceptions.js:113]: I'm happy to debug further if you like.. but it seemed natural to me that `CatchInfo` should only be defined when exception catching is enabled.

**Comment by sbc100** [comment]
Currently depends on two other PRs that need to land first.

File: emcc.py [hunk 2]
```diff
@@ -1401,6 +1401,11 @@ def default_setting(name, new_default):
       # memory init file is not supported with side modules, must be executable synchronously (for dlopen)
       options.memory_init_file = False
 
+    # If we are including the ensure JS library then we know for sure we will by definition
```
Comment (juj) [inline emcc.py:1404]: "If we are including the ensure JS library"? -> "If we are including the full JS library"?

**Comment by juj** [comment]
Very nice!

I wonder how close we would be to enabling Closure warnings by default. About two years ago I worked a lot to get there, but there were some things that were too much work to clean up, maybe things might have improved since. (although recently when I looked,  I did see some new regressions..)

File: src/library_exceptions.js (hunk 1 above)
Comment (kripken) [reply src/library_exceptions.js]: Ah - I think the issue is that the file gets included at all. `modules.js` will not include this file if exceptions are disabled, so these ifdefs should be redundant. Is the test perhaps using INCLUDE_FULL_LIBRARY? Then perhaps we need to make that not include files that do not make sense (like this file without exceptions).

Comment (sbc100) [reply src/library_exceptions.js]: From modules.js:

```
     if (!DISABLE_EXCEPTION_THROWING) {                                                                                      
        libraries.push('library_exceptions.js');                                 
      } else {                                                                   
        libraries.push('library_exceptions_stub.js');                            
      }     
```

It looks like only DISABLE_EXCEPTION_THROWING=1 will cause the stubs to be included.. DISABLE_EXCEPTION_CATCHING=1 (which is the default) does not use the stubs.  Instead it includes this file.

Comment (sbc100) [reply src/library_exceptions.js]: Should `DISABLE_EXCEPTION_CATCHING=1` have its own stubs file?

Comment (sbc100) [reply src/library_exceptions.js]: I spit this part of the change into https://github.com/emscripten-core/emscripten/pull/13862 so we can continue the discussion there.

File: src/library_webgl.js [hunk 3]
```diff
@@ -3536,8 +3536,10 @@ var LibraryGL = {
   glBindVertexArrayOES: 'glBindVertexArray',
   glIsVertexArrayOES: 'glIsVertexArray',
 
+#if LEGACY_GL_EMULATION
```
Comment (kripken) [inline src/library_webgl.js]: these functions can be moved to the separate `library_glemu.js`

**Review by kripken — APPROVED** [review]

File: src/library_webgl.js (hunk 3 above)
Comment (sbc100) [reply src/library_webgl.js]: The could.. but I wasn't sure that the rule was here.. there seem to be other usages of LEGACY_GL_EMULATION in this file.   I will propose that as followup.

Comment (sbc100) [reply src/library_webgl.js]: Actually you are right.  I moved them as part of this PR.

**Pull request merged**